src/WebJobs.Extensions.DurableTask/Listener/OutOfProcOrchestrationShim.cs

```diff
@@ -104,6 +125,195 @@ internal async Task<bool> ScheduleDurableTaskEvents(OrchestrationInvocationResul
+                        var entityId = EntityId.GetEntityIdFromSchedulerId(action.InstanceId);
+                        this.context.SignalEntity(entityId, action.FireAt, action.EntityOperation, action.Input);
+                        break;
+                    }
+
+                case AsyncActionType.ContinueAsNew:
+                    this.context.ContinueAsNew(action.Input);
+                    break;
+                case AsyncActionType.WaitForExternalEvent:
+                    task = this.context.WaitForExternalEvent<object>(action.ExternalEventName);
+                    break;
+                case AsyncActionType.CallHttp:
+                    task = this.context.CallHttpAsync(action.HttpRequest);
+                    break;
+                default:
+                    break;
+            }
+
+            return task;
+        }
+
+        /// <summary>
+        /// Replays the orchestration execution from an OOProc SDK in .NET.
+        /// This is the V2 implementation of that replay procedure.
+        /// </summary>
+        /// <param name="actionSet">The OOProc actions payload.</param>
+        /// <returns>An awaitable Task that completes once replay completes.</returns>
+        private async Task ProcessAsyncActionsV2(AsyncAction[][] actionSet)
+        {
+            foreach (AsyncAction[] actions in actionSet)
+            {
+                Task newTask;
+                AsyncAction topLevelAction = actions[0];
+
+                if (actions.Length == 1)
+                {
+                    newTask = this.InvokeAPIFromAction(topLevelAction);
+                }
+                else
+                {
+                    // If there's more than one action, we know we're looking at a compound Task / a task of tasks
+                    // such as Task.WhenAll or Task.WhenAny. Recursively, we construct that compound task.
+                    (newTask, _) = this.GetCompoundTaskFromActions(actions, 0, topLevelAction.Depth, topLevelAction.TaskGrouping);
+                }
+
+                // check for null, for when an action is fire-and-forget
+                if (newTask != null)
+                {
+                    await newTask;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Recursively construct a compound Task (a task of tasks) from a list of actions.
+        /// </summary>
+        /// <param name="actions">The OOProc actions payload.</param>
+        /// <param name="index">The index at which to start processing the actions.</param>
+        /// <param name="depth">The current task depth, equal to the number of outer / parent tasks.</param>
+        /// <param name="kind">The kind of parent task: WhenAll or WhenAny.</param>
+        /// <returns>A compound task, either: WhenAll or WhenAny.</returns>
+        private (Task, int) GetCompoundTaskFromActions(AsyncAction[] actions, int index, int depth, TaskGroupingType kind)
```

Comment:
Overall, I am struggling a bit to understand the algorithm without a concrete set of examples of walking through the algorithm and the expected output, along with the algorithm on the SDK side of how orchestrations would generate an AsyncAction[] of this format.

It seems to me that a simpler implementation for both the SDKs and the extension side would be to embed the task structure directly in the data than a flat list with metadata.

For instance, instead of each compound action being an AsyncAction[] that needs to be recursively built into a single Task, why don't we have a single `AsyncAction` with all of the current atomic types that we return, along with two new types `AnyAction` and `AllAction` which have an additional field which is an `AsyncAction[]`. If we are already having a new return schema type, this seems far simpler to both construct on the SDK side, as well as process on the extension side.

Reply:
I see what you mean, and I did think about it 😄 . 
The idea you suggest, with new Action types for `WhenAny` and `WhenAll,` I believe is a better approach if we didn't have to worry about backwards compatibility.

I ended up going against this idea because introducing new Action types would require the OOProc SDKs to know what version of the durable-extension they're communicating with, and I didn't know how to _easily_ have the extension communicate the OOProc schemas it supports.

The approach I prototyped does not require the OOProc SDK to know which OOProc schemas are supported by the extension, the approach is fully backwards compatible with pre-existing logic as we're just adding metadata.